Client/SafeBrowsing/GSafeBrowsingDatabase.swift

```diff
@@ -0,0 +1,509 @@
+    private var numberOfFindRetries: Int16 = 0
+    
+    private var updateTimer: Timer?
+    
+    init() {
+//        self.destroy()
+//        self.persistentContainer = self.create()
+        do {
+            let databaseInfoRequest: NSFetchRequest<ThreatDatabaseInfo> = ThreatDatabaseInfo.fetchRequest()
+            databaseInfoRequest.fetchLimit = 1
+            
+            let databaseInfo = try self.mainContext.fetch(databaseInfoRequest)
+            self.lastFetchDate = databaseInfo.first?.lastFetchDate ?? Date.distantPast
+            self.lastFetchWaitDuration = databaseInfo.first?.fetchWaitDuration ?? 0.0
+            self.numberOfFetchRetries = databaseInfo.first?.numberOfFetchRetries ?? 0
+            
+            self.lastFindDate = databaseInfo.first?.lastFindDate ?? Date.distantPast
+            self.lastFindWaitDuration = databaseInfo.first?.findWaitDuration ?? 0.0
+            self.numberOfFindRetries = databaseInfo.first?.numberOfFindRetries ?? 0
+            
+            let request: NSFetchRequest<Threat> = Threat.fetchRequest()
+            let threats = try self.mainContext.fetch(request)
+            try threats.forEach({
+                let hashes = ($0.hashes ?? []).compactMap({ ($0 as? ThreatHash)?.hashData })
+                //hashes.sort(by: { $0.lexicographicallyPrecedes($1) })
+
+                if let checksum = $0.checksum {
+                    if !self.validate(checksum, hashes) {
+                        throw SafeBrowsingError("Database Corrupted")
+                    }
+                }
+            })
+            self.mainContext.reset()
+        } catch {
+            //Remove everything and re-create the database
+            self.destroy()
+            self.persistentContainer = self.create()
+
+            log.error("Safe-Browsing: \(error)")
+        }
+    }
+    
+    private lazy var persistentContainer: NSPersistentContainer = {
+        return self.create()
+    }()
+    
+    private lazy var backgroundContext: NSManagedObjectContext = {
+        let context = persistentContainer.newBackgroundContext()
+        context.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+        return context
+    }()
+    
+    private lazy var mainContext: NSManagedObjectContext = {
+        let context = persistentContainer.viewContext
+        context.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+        context.automaticallyMergesChangesFromParent = true
+        return context
+    }()
+    
+    private func save() {
+        saveContext(mainContext)
+    }
+    
+    private func saveContext(_ context: NSManagedObjectContext) {
+        if context.hasChanges {
+            do {
+                try context.save()
+            } catch {
+                log.error("Error Saving Context: \(error)")
+            }
+        }
+    }
+    
+    func canFind() -> Bool {
+        return Date().timeIntervalSince(self.lastFindDate) >= self.lastFindWaitDuration
```

Comment:
`lastFindWaitDuration` is in minutes, but `Date().timeIntervalSince()` is in seconds I think.

Reply:
Fixed.